File: financeos/wire/validator_transport.py
"""The FinanceOS_Response_Validator boundary on the wire, Python side (task 29.3).

Requirements 15.1, 15.8; supports 11.11 and 12.6.

The mirror of ``src/wire/validator-transport.ts``. The second of design.md's "three
places money crosses", and the one where the Python side does the comparing: the
Validator is Python, so ``allowed_values_paise`` arrives as decimal strings and is
parsed to ``int`` **before any set-membership comparison**.

:meth:`ResponseValidatorRequestWire.allowed_values` is that parse, and it returns a
``frozenset[int]`` rather than a list because the only question asked of the set is
membership. Requirement 11.11 is a zero-tolerance exact match, which is meaningful
only if the set arrived exactly: a coerced double anywhere in it would silently
widen what counts as grounded — an invented figure matching a neighbouring rounded
member — or narrow it, withholding a response that was correct. The set would have
the right length, the right shape and the wrong contents.
"""

from typing import Final, Literal

from pydantic import Field

from financeos.wire.fields import (
    NullablePaiseWireField,
    PaiseWireField,
    UuidField,
    WireModel,
)
from financeos.wire.paise import from_wire

#: Requirement 11.9's payload bound: at most 200 tool values reach the Model, so at
#: most 200 values can be grounded. A larger allowed set would describe figures the
#: Model was never given.
MAX_ALLOWED_VALUES: Final[int] = 200

#: Requirement 11.9's input ceiling, and therefore the narrative ceiling.
MAX_NARRATIVE_CHARS: Final[int] = 100_000

#: Requirement 11.10 truncates Model output at 8000 characters before release.
MAX_RELEASED_CHARS: Final[int] = 8_000


class ResponseValidatorRequestWire(WireModel):
    """What an Agent hands the Validator.

    No ``tenant_id``: design.md's in-process ``ResponseValidator.validate`` shows
    one, but on the wire the Tenant comes from the forwarded session and a
    body-supplied one is rejected as a schema violation (Requirement 12.7, 14.8).
    """

    narrative: str = Field(min_length=1, max_length=MAX_NARRATIVE_CHARS)
    allowed_values_paise: list[PaiseWireField] = Field(max_length=MAX_ALLOWED_VALUES)
    evidence_chain_ids: list[UuidField] = Field(max_length=MAX_ALLOWED_VALUES)

    def allowed_values(self) -> frozenset[int]:
        """The allowed set as ``int``, parsed before any comparison (Requirement 11.11)."""
        return frozenset(
            from_wire(value, "allowed_values_paise") for value in self.allowed_values_paise
        )


class ValidationReleasedWire(WireModel):
    """Every figure matched an allowed value and every chain resolved."""

    ok: Literal[True]
    released: str = Field(max_length=MAX_RELEASED_CHARS)


class UngroundedFigureWire(WireModel):
    """A monetary token in the narrative matched no allowed value. The whole response
    is withheld (Requirement 11.11).

    ``parsed_paise`` is nullable because a token that could not be normalised to
    paise at all still has to be reported: the figure text is what a reviewer needs,
    and ``None`` states that no paise value was recoverable from it. When a value
    *was* recovered it is reported at the precision it was compared at.
    """

    ok: Literal[False]
    kind: Literal["ungrounded_figure"]
    figure_text: str = Field(min_length=1, max_length=200)
    parsed_paise: NullablePaiseWireField

    def parsed(self) -> int | None:
        """The offending figure as an ``int``, or ``None`` when none was recoverable."""
        if self.parsed_paise is None:
            return None
        return from_wire(self.parsed_paise, "parsed_paise")


class UnresolvedEvidenceChainWire(WireModel):
    """A cited Evidence_Chain identifier did not resolve (Requirement 12.6)."""

    ok: Literal[False]
    kind: Literal["unresolved_evidence_chain"]
    evidence_chain_id: UuidField


ValidationResultWire = ValidationReleasedWire | UngroundedFigureWire | UnresolvedEvidenceChainWire

File: src/wire/validator-transport.ts
/**
 * The FinanceOS_Response_Validator boundary on the wire (task 29.3).
 * Requirements 15.1, 15.8; supports 11.11 and 12.6.
 *
 * The second of design.md's "three places money crosses". The Validator is Python,
 * so `allowed_values_paise` leaves TypeScript as a list of decimal strings and is
 * parsed to `int` on the far side **before any set-membership comparison**.
 *
 * ## Why this list in particular must not be coerced
 *
 * Requirement 11.11 is a zero-tolerance exact match: every monetary token in a
 * Model narrative must equal a member of `allowed_values_paise` exactly, or the
 * whole response is withheld. That test is only meaningful if the allowed set
 * arrived exactly. A coerced double anywhere in the set would silently widen what
 * counts as grounded — a figure the Model invented would match a neighbouring
 * rounded member — or narrow it, withholding a response that was correct. Either
 * way the failure is invisible: the set has the right length, the right shape and
 * the wrong contents.
 *
 * ## `tenant_id` is absent, deliberately
 *
 * design.md's `ResponseValidator.validate` shows a `tenant_id` argument because it
 * describes the in-process interface. On the wire the Tenant comes from the
 * forwarded session and a body-supplied one is rejected as a schema violation
 * rather than ignored (Requirement 12.7, 14.8), so the request schema does not
 * declare it and `z.strictObject` names it if a caller sends it anyway.
 */

import { z } from 'zod';

import { nullablePaiseWire, paiseWireArray } from './paise-schema';

/**
 * Requirement 11.9's payload bound: at most 200 tool values reach the Model, so at
 * most 200 values can be grounded. A larger allowed set would describe figures the
 * Model was never given.
 */
export const MAX_ALLOWED_VALUES = 200;

/** Requirement 11.9's input ceiling, and therefore the narrative ceiling. */
export const MAX_NARRATIVE_CHARS = 100_000;

/** Requirement 11.10 truncates Model output at 8000 characters before release. */
export const MAX_RELEASED_CHARS = 8_000;

/**
 * What an Agent hands the Validator.
 *
 * `allowed_values_paise` is exactly the Financial_Tool outputs supplied to the
 * Model — not a superset, and not a set assembled after the fact.
 */
export const responseValidatorRequestWire = z.strictObject({
  narrative: z.string().min(1).max(MAX_NARRATIVE_CHARS),
  allowed_values_paise: paiseWireArray({ max: MAX_ALLOWED_VALUES }),
  evidence_chain_ids: z.array(z.uuid()).max(MAX_ALLOWED_VALUES),
});

/**
 * design.md's `ValidationResult` on the wire.
 *
 * `parsed_paise` is nullable because a token that could not be normalised to paise
 * at all still has to be reported — the figure text is what a reviewer needs, and
 * `null` states that no paise value was recoverable from it. It is a decimal string
 * whenever it is present, so the withheld figure is reported at the same precision
 * it was compared at.
 */
export const validationResultWire = z.union([
  z.strictObject({ ok: z.literal(true), released: z.string().max(MAX_RELEASED_CHARS) }),
  z.strictObject({
    ok: z.literal(false),
    kind: z.literal('ungrounded_figure'),
    figure_text: z.string().min(1).max(200),
    parsed_paise: nullablePaiseWire,
  }),
  z.strictObject({
    ok: z.literal(false),
    kind: z.literal('unresolved_evidence_chain'),
    evidence_chain_id: z.uuid(),
  }),
]);

export type ResponseValidatorRequestWire = z.infer<typeof responseValidatorRequestWire>;
export type ValidationResultWire = z.infer<typeof validationResultWire>;
